crates/interpreter/src/interpreter/shared_stack.rs

```diff
@@ -0,0 +1,710 @@
+use crate::{
+    primitives::{B256, U256},
+    InstructionResult,
+};
+use alloc::vec;
+use alloc::vec::Vec;
+use core::fmt;
+
+/// EVM interpreter stack limit.
+pub(crate) const STACK_LIMIT: usize = 1024;
+
+const PAGE_SIZE: usize = 4 * STACK_LIMIT;
+type Buffer = Vec<U256>;
+type Checkpoints = Vec<usize>;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+struct Page {
+    /// The underlying buffers shared between calls
+    buffer: Buffer,
+    /// Stack checkpoints for each depth
+    ///
+    /// Invariant: these are always in bounds of `buffer`.
+    checkpoints: Checkpoints,
+}
+
+impl Page {
+    #[inline]
+    fn new() -> Self {
+        let checkpoints = Vec::with_capacity(32);
+        Self {
+            buffer: Vec::with_capacity(PAGE_SIZE),
+            checkpoints,
+        }
+    }
+}
+
+/// A sequential stack shared between calls, which uses
+/// a vector of "pages" (buffers) or internal representation.
+///
+/// Each page includes a buffer of size `PAGE_SIZE` which
+/// is shared between calls, and when there is no more space left
+/// we move to a new page.
+///
+/// A [SharedStack] instance should always be obtained using
+/// the `new` static method to ensure memory safety.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SharedStack {
+    /// The underlying data divided in pages,
+    /// where each page has size `PAGE_SIZE`
+    pages: Vec<Page>,
+    /// Keeps track of the page used for the current context
+    /// Invariant: is a valid index of `self.pages`
+    page_idx: usize,
+    /// The pointer to the buffer used by the current page.
+    /// Needed for better performance and to avoid double
+    /// heap lookup for basic stack operations.
+    ///
+    /// Invariant: it a valid pointer to `self.pages[self.page_idx].buffer`
+    buffer: *mut Buffer,
```

Comment:
Would put full `Page` here not just buffer. and rename `pages` field to `previous_pages` to represent pages in somekind of backlog (not currently active)

Reply:
hm i see the problem here, maybe we should not look at free pages as a stack. wdyt about idea of having two fields: `taken_pages` and `free_pages` and they are both Vecs.

 if the context is freed you put the Page to the `free_pages` vec and on `new_context` you pop page from `free_pages` or create a new one if there is none.

Reply:
Another idea (not related to the first part) is related to the `Buffer` and the `context_len`, Buffer is a `Vec` here that has its own `len` and additionally we have `context_len` that we increment. 

We could work with `*mut U256` which would point us directly to the top of the stack, and we could have `context_len` that would tell us the bound of the stack. I like this idea, but we should be careful about the usage of the pointer.

Reply:
I like the first idea! It should make it easier to reason about and we can avoid the `page_idx` which can be annoying to work with.

Regarding the second idea let me know if I got this right: with a raw pointer to the top of the stack, all operations are not done on the buffer itself (which is preallocated with 1024 capacity) but rather on the pointer, by dereferencing and incrementing/decrementing it. If that's the case, do you think it provides a performance improvement over dealing with the `Buffer` pointer or it is a matter of ergonomics?

Lastly, but slighly unrelated: wdyt of putting both shared stack and shared memory under the `EVMContext` struct? Then when creating the interpreter we can pass a raw pointer to the current context buffer of stack and memory.

Reply:
Usage of the pointer should be slightly faster, it should be less ergonomical as you directly handle the pointer.

Would be good to put it in `EvmContext` but I think borrowing is going to kill us there, and it would not matter a lot.

Reply:
Okay, I'll try to modify the shared stack with suggestions. It will be some work to do, I hope some wait is not a problem!

Regarding `EVMContext` maybe I can think of it in a separate PR after this and see if it is reasonable

Reply:
Take your time and pace yourself, I am fine to take over if you feel burdened to finish it, and I am fine with waiting. Both things work for me.

We can maybe saparate that pointer idea to another PR to not clog this.

Reply:
Yeah if you want I'd be very happy to work on the pointer idea on a separate PR!